libafl/src/executors/inprocess/mod.rs

```diff
@@ -41,150 +41,162 @@ pub mod inner;
+
+/// The inmem executor simply calls a target function, then returns afterward.
 #[allow(dead_code)]
-pub struct GenericInProcessExecutor<H, HB, HT, OT, S>
-where
-    H: FnMut(&S::Input) -> ExitKind + ?Sized,
-    HB: BorrowMut<H>,
-    HT: ExecutorHooksTuple<S>,
-    OT: ObserversTuple<S>,
-    S: State,
-{
-    harness_fn: HB,
-    inner: GenericInProcessExecutorInner<HT, OT, S>,
-    phantom: PhantomData<(*const H, HB)>,
+pub struct GenericInProcessExecutor<H, HT, OT> {
+    harness: H,
+    hooks: Option<HT>,
+    observers: OT,
 }
 
-impl<H, HB, HT, OT, S> Debug for GenericInProcessExecutor<H, HB, HT, OT, S>
+impl<H, HT, OT> Debug for GenericInProcessExecutor<H, HT, OT>
 where
-    H: FnMut(&S::Input) -> ExitKind + ?Sized,
-    HB: BorrowMut<H>,
-    HT: ExecutorHooksTuple<S>,
-    OT: ObserversTuple<S> + Debug,
-    S: State,
+    OT: Debug,
 {
     fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
         f.debug_struct("GenericInProcessExecutor")
-            .field("inner", &self.inner)
-            .field("harness_fn", &"<fn>")
+            .field("observers", &self.observers)
             .finish_non_exhaustive()
     }
 }
 
-impl<H, HB, HT, OT, S> UsesState for GenericInProcessExecutor<H, HB, HT, OT, S>
-where
-    H: FnMut(&S::Input) -> ExitKind + ?Sized,
-    HB: BorrowMut<H>,
-    HT: ExecutorHooksTuple<S>,
-    OT: ObserversTuple<S>,
-    S: State,
-{
-    type State = S;
+impl<H, HT, OT> HasObservers for GenericInProcessExecutor<H, HT, OT> {
+    type Observers = OT;
+
+    #[inline]
+    fn observers(&self) -> RefIndexable<&Self::Observers, Self::Observers> {
+        RefIndexable::from(&self.observers)
+    }
+
+    #[inline]
+    fn observers_mut(&mut self) -> RefIndexable<&mut Self::Observers, Self::Observers> {
+        RefIndexable::from(&mut self.observers)
+    }
 }
 
-impl<H, HB, HT, OT, S> UsesObservers for GenericInProcessExecutor<H, HB, HT, OT, S>
-where
-    H: FnMut(&S::Input) -> ExitKind + ?Sized,
-    HB: BorrowMut<H>,
-    HT: ExecutorHooksTuple<S>,
-    OT: ObserversTuple<S>,
-    S: State,
-{
-    type Observers = OT;
+#[derive(Debug)]
+pub struct RecoveryHook;
+
+impl<E, EM, I, S, Z> ExecutorHook<E, EM, I, S, Z> for RecoveryHook {
+    fn pre_exec(
+        &mut self,
+        executor: &mut E,
+        fuzzer: &mut Z,
+        state: &mut S,
+        mgr: &mut EM,
+        input: &S,
+    ) {
+        todo!()
+    }
+
+    fn post_exec(
+        &mut self,
+        executor: &mut E,
+        fuzzer: &mut Z,
+        state: &mut S,
+        mgr: &mut EM,
+        input: &S,
+    ) {
+        todo!()
+    }
 }
 
-impl<EM, H, HB, HT, OT, S, Z> Executor<EM, Z> for GenericInProcessExecutor<H, HB, HT, OT, S>
+/// This function marks the boundary between the fuzzer and the target for in-process executors
+///
+/// # Safety
+/// This function sets a bunch of raw pointers in global variables, reused in other parts of
+/// the code.
+#[inline]
+pub fn enter_target<EM, I, S, Z>(
```

Comment:
This should be unsafe

Reply:
Yup, this is going to disappear soon anyways (reimplementing this entirely in hooks/callbacks)